otter/test/convergence/test_selfheal.py

```diff
@@ -0,0 +1,327 @@
+        """
+        When lock is not acquired, it is tried and if failed does not
+        call _perform
+        """
+        self.ila.return_value = succeed(False)
+        self.lock.acquire.return_value = fail(LockTimeout())
+        # tests that it is not called
+        self.s._perform = lambda: 1 / 0
+        self.s.startService()
+        self.ila.assert_called_once_with(self.s.disp, self.lock)
+        self.lock.acquire.assert_called_once_with(True, 0.1)
+
+    def test_converge_all_lock_acquired(self):
+        """
+        When lock is not acquired, it is tried and if successful self._perform
+        is called
+        """
+        self.ila.return_value = succeed(False)
+        self.lock.acquire.return_value = succeed(True)
+        self.s._perform = mock.Mock()
+        self.s.startService()
+        self.ila.assert_called_once_with(self.s.disp, self.lock)
+        self.lock.acquire.assert_called_once_with(True, 0.1)
+        self.s._perform.assert_called_once_with()
+        self.log.msg.assert_called_once_with(
+            "self-heal-lock-acquired", otter_service="selfheal")
+
+    def test_converge_all_lock_already_acquired(self):
+        """
+        If lock is already acquired, it will just call self._perform
+        """
+        self.s._perform = mock.Mock()
+        self.s.startService()
+        self.ila.assert_called_once_with(self.s.disp, self.lock)
+        # Lock is not acquired again
+        self.assertFalse(self.lock.acquire.called)
+        self.s._perform.assert_called_once_with()
+
+    def test_converge_all_kz_not_connected(self):
+        """
+        If kazoo client is not connected then nothing is done
+        """
+        self.kzc.state = KazooState.LOST
+
+        def bad_func(*a):
+            return 1 / 0
+
+        self.s._perform = self.ila.side_effect = \
+            self.lock.acquire.side_effect = bad_func
+        self.s.startService()
+        self.log.err.assert_called_once_with(
+            mock.ANY, "self-heal-kz-state", state=KazooState.LOST,
+            otter_service="selfheal")
+
+    def test_performs_again(self):
+        """
+        Calls _perform at every interval
+        """
+        self.s._perform = mock.Mock()
+        self.s.startService()
+        self.s._perform.assert_called_once_with()
+        self.clock.advance(300)
+        self.assertEqual(self.s._perform.call_count, 2)
+
+    def test_performs_again_on_err(self):
+        """
+        Calls _perform at every interval even it it fails
+        """
+        self.s._perform = mock.Mock(return_value=fail(ValueError("h")))
+        self.s.startService()
+        self.s._perform.assert_called_once_with()
+        self.log.err.assert_called_once_with(
+            CheckFailure(ValueError), otter_service="selfheal")
+        self.s._perform.return_value = succeed(None)
+        self.clock.advance(300)
+        self.assertEqual(self.s._perform.call_count, 2)
+
+    @mock.patch("otter.convergence.selfheal.check_and_trigger")
```

Comment:
Because the returned effect with have a `Func` intent due to `do` decorator that needs to be resolved to have its callback called to know what parameters `check_and_trigger` was called with. Mock is not used when we are completely in effect world (see tests for `check_and_trigger` and `is_lock_acquired`) but I find it difficult to write tests without mock in mixed case like this one. I mean its doable but gets complicated.